core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,731 @@
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        return ehsm_calc_keyblob_size(cmk->metadata.keyspec, cmk->keybloblen);
+    }
+
+    if (cmk->metadata.keyspec != EH_AES_GCM_128 &&
+        cmk->metadata.keyspec != EH_AES_GCM_192 &&
+        cmk->metadata.keyspec != EH_AES_GCM_256)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint32_t keysize = 0;
+    if (!ehsm_get_symmetric_key_size(cmk->metadata.keyspec, keysize))
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+
+    uint8_t *key = (uint8_t *)malloc(keysize);
+    if (key == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+    ret = sgx_read_rand(key, keysize);
+    if (ret != SGX_SUCCESS)
+    {
+        free(key);
+        return ret;
+    }
+    ret = ehsm_create_keyblob(key,
+                              keysize,
+                              (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    memset_s(key, keysize, 0, keysize);
+
+    free(key);
+    return ret;
+}
+
+#ifdef FIPS_MODE
+static bool test_RSA_keypair(RSA *keypair)
+{
+    uint8_t data2sign[] = "test_rsa_keypair";
+    uint8_t signature[RSA_size(keypair)] = {0};
+    uint32_t data2sign_size = sizeof(data2sign) / sizeof(data2sign[0]);
+    bool result = false;
+    if (rsa_sign(keypair, 
+                 EVP_sha256(), 
+                 EH_PAD_RSA_PKCS1_PSS, 
+                 data2sign, 
+                 data2sign_size, 
+                 signature, 
+                 RSA_size(keypair)) != SGX_SUCCESS)
+        return false;
+
+    if (rsa_verify(keypair, 
+                   EVP_sha256(), 
+                   EH_PAD_RSA_PKCS1_PSS, 
+                   data2sign, 
+                   data2sign_size, 
+                   signature, 
+                   RSA_size(keypair), 
+                   &result, -1) != SGX_SUCCESS)
+        return false;
+    
+    return result;
+}
+#endif
+
+sgx_status_t ehsm_create_rsa_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        return ehsm_calc_keyblob_size(cmk->metadata.keyspec, cmk->keybloblen);
+    }
+
+    RSA *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    uint8_t *pem_keypair = NULL;
+    uint32_t key_size = 0;
+
+    rsa_keypair = RSA_new();
+    if (rsa_keypair == NULL)
+    {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec) // https://github.com/intel/linux-sgx/blob/master/SampleCode/SampleAttestedTLS/common/utility.cpp
+    {
+    case EH_RSA_2048:
+        rsa_keypair = RSA_generate_key(RSA_2048_KEY_BITS, RSA_F4, NULL, NULL);
+        break;
+    case EH_RSA_3072:
+        rsa_keypair = RSA_generate_key(RSA_3072_KEY_BITS, RSA_F4, NULL, NULL);
+        break;
+    case EH_RSA_4096:
+        rsa_keypair = RSA_generate_key(RSA_4096_KEY_BITS, RSA_F4, NULL, NULL);
```

Comment:
RSA_generate_key() is deprecated (new applications should use RSA_generate_key_ex()

https://manpages.ubuntu.com/manpages/bionic/man3/RSA_generate_key_ex.3ssl.html

Reply:
done